### ISSUE TITLE
Add Tailscale-based OIDC IdP server

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#10263]
It'd be nice to be able to log in to Synology (and Proxmox PVE and Proxmox Backup Server and ....) with Tailscale auth, instead of a username and password.

Those products support being an OpenID Connect (OIDC) relying party (client).

Tracking bug to flesh out a OIDC server.

/cc @maisem

**Comment by anderspitman** [comment]
Interesting. So you set up your Synology to use the TS address of a node running tsidp, then you can use OIDC to log in? Doesn't look like there's any UI on the /authorize endpoint for the current implementation, it just auto-redirects back to the RP with a code. I guess since you're already on the tailnet and it appears to return the current user, you don't need to do all the usual security song and dance with user consent?

Any chance of this eventually being available over Funnel? Just a couple days ago [I submitted](https://github.com/tailscale/tailscale/issues/10243) an FR for Tailscale to integrate more seamlessly with self-hosted OIDC servers. If every TS client *became* an OIDC server, that would be a huge boon for decentralized identity. Especially if Funnel eventually supports custom domains.

**Comment by anderspitman** [comment]
Of course then in the case of Tailscale you end up with a chicken-egg problem. You want to use your own OIDC server to log in to TS, but you also want to use TS Funnel to host your OIDC server :laughing:

**Comment by bradfitz** [comment]
Yes, no UI. It works great! It works for Synology and Proxmox Backup Server and Proxmox PVE so far at least. Docs coming.

@maisem was working on Funnel support for non-Tailscale Relying Parties.

**Comment by btrepp** [comment]
This is awesome. I had played around with this but great to see it in the tailscale source. You are already doing a bit of Auth being on the network anyway, and this could have great bolt on security benefits with say oauth2-proxy for older apps.

Oauth2 proxy configure to require you to be logged into the toilet idp to get to the 'legacy' app(s) and you have a good default story on stuff in your tailnet.

For hosting it, I imagine this is self hosted to begin with, but could be great if it's tailscale hosted (similar to how the SSH functionality works). Ideally what I want is just the oidc functionality to go this is an authorised user in your tailnet, so that can be confirmed at the app layer.

Having it run on each tailnet node is a clever as hell idea though.

**Comment by pavanbuzz** [comment]
@bradfitz Thanks for all the great work! This is an awesome feature!

Is it possible to use this feature now, if so is there any docs available? I would love to use this feature with Proxmox and other OIDC RP clients.

**Comment by btrepp** [comment]
I've quickly thrown it in a container https://hub.docker.com/r/btrepp/tsidp

Works quite well. Playing with things like kubenav and kubectl and got it working with k8s.

What would be nice is refresh token support. Kubenav seems to not prompt for reauth and wants to use the refresh token. Having refresh tokens would be excellent

**Comment by madjam002** [comment]
Just tried this out and it's working really well with Hashicorp Vault to automatically log in, would be good to get the app layer grants coming through in the ID token, that way policies could automatically be applied based on ACLs in Tailscale :)

**Comment by florisvdg** [comment]
For a relying party to verify the token claims, will the `.well-known/openid-configuration` and JWKS be available on tailscale.com or does the relying party need to be part of the tailnet for this to work?

**Comment by arunoruto** [comment]
Similar to @btrepp , I created a containerized version here: https://github.com/arunoruto/tsidp
It is built automatically every Sunday against the newest release and available via GitHub Container Repository `ghcr.io/arunoruto/tsidp`.

**Comment by mitchellkellett** [comment]
I gave this a go recently with Tailscale and loved how I could use it to authenticate automatically when using a machine in the Tailnet. I wanted to implement it in my home lab with Headscale, but there's no flag for setting the login server. 😞

**Comment by cedi** [comment]
I absolutely love tsidp. I'm using it in more and more of my infrastructure. It's just perfect for a small home-network environment where I have like 2 or 3 users and I don't have to maintain a whole IDP Server just for a few users to login to my Synology and HomeAssistant.

However, I have one feature request to tsidp: would it be possible to add a `groups` claim?
My thinking goes like this: In my ACL policy file I specify groups and add individual users to groups. So it must be possible somehow to get the groups a user belongs to in the tsidp code.

I tried implementing this today so I could send a PR for this, but I was banging my head against the wall on how to get the ACL policy file in cmd/tsidp/tsidp.go.
The [localapi](https://github.com/tailscale/tailscale/blob/main/ipn/localapi/localapi.go) exposes a `policy/` endpoint via [`GetEffectivePolicy()`](https://github.com/tailscale/tailscale/blob/main/client/local/local.go#L848) but I couldn't make sense of this API endpoint and just couldn't get any meaningful policy back. 
I am reluctant on [/api/v2/tailnet/{tailnet}/acl](https://tailscale.com/api#tag/policyfile/GET/tailnet/{tailnet}/acl) because that requires additional code in tsidp.go to obtain a bearer token and I just feel a bit uncomfortable calling the public API of tailscale from my IDP server "just" to get the ACL policy file so I can eventually evaluate the login-users group memberships. 
The admin portal frontend makes an API call to `https://login.tailscale.com/admin/api/acls/user?user=XXX` which returns a JSON with 

```json
{
    "status": "success",
    "data": {
        "userGroups": [
            "group:group1",
            "group:group2"
        ],
        "userAutoGroups": [
            "autogroup:member",
            "autogroup:owner",
            "autogroup:admin"
        ],
        "userSyncedGroups": null,
        "allGroups": [
            "group:group1",
            "group:group2",
            "group:group3"
        ]
    }
}
```

which is more like what I would be looking for. 

@bradfitz could you advise or give me some pointers how I could obtain the user's group memberships in tsidp.go so I can send a PR to include a groups claim in the JWST token?

**Comment by Vexed01** [comment]
Hi, sorry if this is the wrong place but I'm a little stuck!

I had set up tsidp with Immich and Jellyfin and got both working a couple weeks ago, but a few days ago it suddenly stopped working. Jellyfin's logs aren't very helpful but Immich is complaing about no refresh_token being provided. I've taken a look at the go code (not my are of expertise!) and it looks like these are never generated by tsidp. Is refresh_token being expected an issue anyone else has faced? What

**Comment by bradfitz** [comment]
Closing. https://github.com/tailscale/tsidp/ exists nowadays.
